feat: add `svelte/html-closing-bracket-spacing`

marekvospel opened the pull request:
Added `html-closing-bracket-spacing` rule, which can be seen in [`eslint-plugin-vue`](https://github.com/vuejs/eslint-plugin-vue)

## Related PRs

feat: add `svelte/html-self-closing` #190 
feat: add `svelte/no-spaces-around-equal-signs-in-attribute` #191

## `svelte/html-closing-bracket-spacing`
A simple rule for enforcing a space before closing a tag
```html
<div/>
<!-- gets fixed to -->
<div />
```

## Coverage
100% test coverage.

## ~~Notes & questions~~
* ~~This is my first time contributing to this project and creating eslint plugins in general, so I'm sorry if some parts of my code aren't as good.~~
* ~~How should I document the rules? Is there a script to automatically generate the documentation, or should I create a new markdown file inside `docs/rules`? (This information should probably be added to contributing section in README)~~

ota-meshi commented:
Thank you for this PR!
You need to add documents that describes the rules you added. After adding the files, run `yarn update` to update the header and footer.

marekvospel commented:
@ota-meshi I've added the docs, fixed the merge conflicts and both the docs build and lint should be passing now!

ota-meshi commented:
Thank you for adding the documentation! I'm busy this week so I'll probably check it next week.

marekvospel commented:
I had extra time today, so I have added `svelte/html-closing-bracket-spacing` rule!
```html
<div/>
# gets fixed to
<div />
```
It's similar to to `svelte/no-spaces-around-equal-signs-in-attribute `, but only has to check the end of the tag, so it's a bit simpler rule.

The test also has 100% coverage

I've also fixed some minor problems in my docs (such as since version being 0.2.2, which is already released, or mistyped `<div />`

ota-meshi commented on docs/rules/html-self-closing.md:
```diff
@@ -0,0 +1,82 @@
+      "html": {
+        "void": "never", // or "always" or "any"
+        "normal": "always", // or "never" or "any"
+        "custom": "always" // or "never" or "any"
```
```suggestion
        "component": "always" // or "never" or "any"
```

ota-meshi commented on docs/rules/html-self-closing.md:
```diff
@@ -0,0 +1,82 @@
+
+- `html.void` (`"never"` by default)... Style of HTML void elements
+- `html.component` (`"always"` by default)... Style of svelte components
+- `html.normal` (`"always"` by default)... Style of other elements
```
`<svelte:self></svelte:self>` seems to be affected by the `normal` option, but I think they need to share the `component` option or have different option.

marekvospel replied:
I've moved this to #190, as it is related to `svelte/html-self-closing` rule and might take me longer to resolve.

ota-meshi commented on docs/rules/html-self-closing.md:
````diff
@@ -0,0 +1,82 @@
+}
+```
+
+- `html.void` (`"never"` by default)... Style of HTML void elements
````
Could you make `"always"` the default to be the same as the default for prettier-plugin-svelte?

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+export default createRule("html-closing-bracket-spacing", {
+  meta: {
+    docs: {
+      description: "Require or disallow a space before tag's closing brackets",
```
The first letter should be lowercase to be consistent with existing rules.

```suggestion
      description: "require or disallow a space before tag's closing brackets",
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+        const tagType =
+          node.type === "SvelteEndTag"
+            ? "endTag"
+            : selfClosing
```
I think you can use `node.selfClosing`.

ota-meshi replied:
Hmmm.  I checked on the parser demo site and each seems to return the following results:

```html
<img /> <!-- "selfClosing": true -->
<img> <!-- "selfClosing": false -->
```

https://ota-meshi.github.io/svelte-eslint-parser/

Can you share the template you tried, as there may be other causes?

marekvospel replied:
I'm not sure rn, I'll try to find what caused my issue

marekvospel replied:
Replacing "/" check with `node.selfClosing` doesn't break any tests, and I couldn't find any edge case where `startTag.selfClosing` would be `false` even if it was a self closing tag, so I think it was me mistaking something while adding this rule.

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+      "SvelteStartTag, SvelteEndTag"(
+        node: AST.SvelteStartTag | AST.SvelteEndTag,
+      ) {
+        if (node.type === "SvelteEndTag" && options.endTag === "any") return
```
I think we can use `node.selfClosing` to merge the judges with the L111.

```js
        const tagType =
          node.type === "SvelteEndTag"
            ? "endTag"
            : node.selfClosing
            ? "selfClosingTag"
            : "startTag"

        if (options[tagType] === "any") return
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+     * Returns true if string contains newline characters
+     */
+    function containsNewline(string: string): boolean {
+      return /.*\n.*/s.test(string)
```
```suggestion
      return string.includes("\n")
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+        const match = /(\s*)\/?>$/m.exec(tagSrc)
+        if (containsNewline(match![1])) return
+
+        if (options[tagType] === "always" && !containsWhitespace(match![1])) {
```
I don't think we need to use containsWhitespace.

```suggestion
        if (options[tagType] === "always" && !(match![1])) {
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+          report(node, true)
+        } else if (
+          options[tagType] === "never" &&
+          containsWhitespace(match![1])
```
```suggestion
          match![1]
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+      shouldHave: boolean,
+    ) {
+      const tagSrc = src.getText(node)
+      const match = /(\s*)\/?>$/m.exec(tagSrc)
```
Probably don't need the m flag.

```suggestion
      const match = /(\s*)\/?>$/.exec(tagSrc)
```

I think the following HTML will make the wrong decision.

```html
<a attr="
  >
">
```

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,127 @@
+
+        if (node.type === "SvelteStartTag" && options[tagType] === "any") return
+
+        const match = /(\s*)\/?>$/m.exec(tagSrc)
```
```suggestion
        const match = /(\s*)\/?>$/.exec(tagSrc)
```

ota-meshi reviewed (changes requested):
Thank you for this PR!

I have change requests.
I've looked at the source code for one rule, but I'll check the rest after tomorrow.

ota-meshi commented:
If possible, can you separate the PR for each rule?
That way I can split the time to check the PR.

ota-meshi commented:
If you want to create new rules besides these, you may find it useful to use `yarn new new-rule-name`.

JounQin commented on docs-svelte-kit/build-system/src/eslint.mjs:
```diff
@@ -1,6 +1,6 @@
 /* eslint require-jsdoc:0 -- shim */
 
-import * as all from "../../../node_modules/eslint/lib/linter/linter.js"
+import * as all from "../../../node_modules/eslint/lib/linter/linter"
```
Adding `.js` extensions would help to migrate to native ESM later, but of course it should be consistent all through the whole code base.

marekvospel replied:
mb, I've added this when linting the project for some reason 😄

ota-meshi replied:
Yes. It seems that the linting config was broken once. I think it's okay to put it back now.

ota-meshi commented on tests/fixtures/rules/html-end-tags/valid/test01-input.svelte:
```diff
@@ -0,0 +1 @@
+<p>This should be valid</p>
```
It looks like an unnecessary file.

JounQin commented on .gitignore:
```diff
@@ -108,3 +108,6 @@ dist
 /docs-svelte-kit/shim/assert.mjs
+
+# Intellij IDES
+.idea
```
Personally speaking, `IDE` things should be ignored with global `gitignore` file on your own machine.

JounQin commented on tsconfig.json:
```diff
@@ -18,6 +18,7 @@
     //     "*": ["typings/*"]
     // },
     "esModuleInterop": true,
+    "resolveJsonModule": true,
```
This could be broken for ESM later (although I know it is transformed as commonjs), but it could be better to not adding such unstandard config.

JounQin commented:
Hi @marekvospel Can you update this PR’s description align to its actual content?

ota-meshi commented on src/rules/html-closing-bracket-spacing.ts:
```diff
@@ -0,0 +1,109 @@
+      expectedSpace: "Expected space before '>', but not found.",
+      unexpectedSpace: "Expected no space before '>', but found.",
+    },
+    fixable: "code",
```
I think we can use `"whitespace"` because it only auto-fixes whitespace.

ota-meshi reviewed (changes requested):
Thank you for changing this PR. Also, thank you for splitting the PR.
I commented only one. Other than that, it looks good to me.

JounQin commented:
Personally thinking, `styling` rules should be extracted to another plugin like `eslint-plugin-svelte-styling` just like ESLint core has frozen their styling rules. We have prettier for `.svelte` or `.vue` files, although someone may not be using `prettier`, but another `eslint-plugin-svelte-styling` can help, so that `styling` rules can be excluded from this one.

ota-meshi commented:
Hmm. I'm not sure if the stylistic rules should be an another plugin.
For now, I don't think it's necessary. Because the stylistic rules of the ESLint core are not yet another plugin.
But later my thinking may change.

JounQin commented:
> Because the stylistic rules of the ESLint core are not yet another plugin.

ESLint core has frozen their styling rules, no new related rule will be added.

ota-meshi commented:
I know that, but plugin for Svelte doesn't freeze the stylistic rules.
The ESLint core stylistic rules were used by many users and enhanced by many contributors until they froze. However, this plugin is not yet known to users and there are not enough stylistic rules.

ota-meshi reviewed (approved):
LGTM!
